[WIP] jsonEfication

owlishDeveloper opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=1335915

I decided to start with a WIP PR as opposed to asking questions. This is how I currently imagine the place of the `json-e` library in life, Universe and everything. I hope that's correct? What puzzles me is passing tests 😄

I guess I also need to add some tests for the version 1 YAML.

djmitche commented on src/intree.js:
```diff
@@ -103,14 +104,15 @@ module.exports.setup = function(cfg) {
+    //   'taskcluster.docker.provisionerId': cfg.intree.provisionerId,
+    //   'taskcluster.docker.workerType': cfg.intree.workerType,
+    // }));
+    config = jsone(config);
```
jsone takes two inputs -- template and context.  What's in that context is the hard part.  You can pattern it after what we do in mozilla-taskcluster -- https://docs.taskcluster.net/reference/integrations/mozilla-taskcluster/docs/taskcluster-yml

Note that we will also need to render the template on the master branch, without any pushes, to look for `allowPullRequests` and other security-related configuration.  Maybe that can have a different `tasks_for`, perhaps `access_check`?

owlishDeveloper replied:
Thank you Dustin! This is very helpful. I actually tried taking an initial stub at context, and got tangled :) I'll try hack on the context some time this week.

djmitche commented on src/intree.js:
```diff
@@ -89,7 +90,7 @@ function completeInTreeConfig(config, payload) {
   return function({config, payload, validator, schema}) {
     config = yaml.safeLoad(config);
-    let errors = validator(config, schema);
+    let errors = validator(config, schema); // TODO: add version 1 to the schema in base repo
```
..and to the docs :)

owlishDeveloper committed:
Created v1 fixtures

owlishDeveloper committed:
Added a schema

owlishDeveloper committed:
Added another schema

owlishDeveloper committed:
Deleted schema 😜

owlishDeveloper committed:
Corrected the version

owlishDeveloper committed:
Made old tests pass after rebase

owlishDeveloper committed:
Set up a stage, test fails

owlishDeveloper committed:
Started implementation over
Added jsone dependency;
primed a testing canvas;
sorted out validation;
created and deadline are in the template for now;
started adding gh event API into the picture

owlishDeveloper commented:
Decided to push my work so far just to demonstrate I'm alive - there's nothing to review yet, I don't think.

owlishDeveloper committed:
added push event fixture

owlishDeveloper committed:
Got the parameter substitution

owlishDeveloper committed:
Switched to webhook

owlishDeveloper committed:
The push test passes

owlishDeveloper committed:
Some changes. Tests failing

owlishDeveloper committed:
Fixed api test

owlishDeveloper committed:
Intree tests pass (except one)

owlishDeveloper committed:
Adding branch to the api message

owlishDeveloper committed:
All v1 tests pass

owlishDeveloper commented:
https://github.com/taskcluster/taskcluster-github/pull/247
